File: java/src/test/java/mrk/hackerrank/HackerRank08Test.java
package mrk.hackerrank;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class HackerRank08Test {

    private static Stream<Arguments> testData() {
        return Stream.of(
                Arguments.of(new int[]{73, 67, 38, 33}, new int[]{75, 67, 40, 33}),
                Arguments.of(new int[]{37}, new int[]{37}),
                Arguments.of(new int[]{34}, new int[]{34}),
                Arguments.of(new int[]{38}, new int[]{40}),
                Arguments.of(new int[]{40}, new int[]{40}),
                Arguments.of(new int[]{84}, new int[]{85}),
                Arguments.of(new int[]{83}, new int[]{85}),
                Arguments.of(new int[]{82}, new int[]{82}),
                Arguments.of(new int[]{100}, new int[]{100})
        );
    }

    @ParameterizedTest
    @MethodSource("testData")
    void test(int[] grades, int[] expected) {
        Assertions.assertArrayEquals(expected, new HackerRank08().gradingStudents(grades));
    }

    @org.junit.jupiter.api.Test
    void emptyArrayThrows() {
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> new HackerRank08().gradingStudents(new int[]{}));
    }

    @org.junit.jupiter.api.Test
    void tooLargeArrayThrows() {
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> new HackerRank08().gradingStudents(new int[61]));
    }

    @org.junit.jupiter.api.Test
    void gradeAbove100Throws() {
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> new HackerRank08().gradingStudents(new int[]{101}));
    }

    @org.junit.jupiter.api.Test
    void gradeBelow0Throws() {
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> new HackerRank08().gradingStudents(new int[]{-1}));
    }
}

File: kotlin/src/test/kotlin/HackerRank08Test.kt
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class HackerRank08Test {
    @ParameterizedTest(name = "{0}")
    @MethodSource("hackerRank08Data")
    fun test_hackerRank08(
        name: String,
        input: IntArray,
        expected: IntArray,
    ) {
        assertArrayEquals(expected, hackerRank08(input))
    }

    @Test
    fun emptyArrayThrows() {
        assertThrows(IllegalArgumentException::class.java) { hackerRank08(intArrayOf()) }
    }

    @Test
    fun tooLargeArrayThrows() {
        assertThrows(IllegalArgumentException::class.java) { hackerRank08(IntArray(61)) }
    }

    @Test
    fun gradeAbove100Throws() {
        assertThrows(IllegalArgumentException::class.java) { hackerRank08(intArrayOf(101)) }
    }

    @Test
    fun gradeBelow0Throws() {
        assertThrows(IllegalArgumentException::class.java) { hackerRank08(intArrayOf(-1)) }
    }

    companion object {
        @JvmStatic
        fun hackerRank08Data() =
            listOf(
                Arguments.of("example", intArrayOf(73, 67, 38, 33), intArrayOf(75, 67, 40, 33)),
                Arguments.of("no rounding below threshold", intArrayOf(37), intArrayOf(37)),
                Arguments.of("no rounding well below threshold", intArrayOf(34), intArrayOf(34)),
                Arguments.of("rounds up at threshold", intArrayOf(38), intArrayOf(40)),
                Arguments.of("already multiple of 5", intArrayOf(40), intArrayOf(40)),
                Arguments.of("rounds up diff 1", intArrayOf(84), intArrayOf(85)),
                Arguments.of("rounds up diff 2", intArrayOf(83), intArrayOf(85)),
                Arguments.of("no rounding diff 3", intArrayOf(82), intArrayOf(82)),
                Arguments.of("max grade", intArrayOf(100), intArrayOf(100)),
            )
    }
}
